src/vvwe.rs

```diff
@@ -0,0 +1,303 @@
+
+use crate::{Actor, Dot};
+
+/// Version Vector with Exceptions
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CausalityBarrier<A: Actor, T: CausalOp<A>> {
+    peers: HashMap<A, VectorEntry>,
+    // TODO: this dot here keying the T comes from `T::happens_after()`
+    //       Why do we need to store this,
+    pub buffer: HashMap<Dot<A>, T>,
+}
+
+type LogTime = u64;
+
+#[derive(Serialize, Deserialize, Clone, Debug, Default)]
+pub struct VectorEntry {
+    // The version of the next message we'd like to see
+    next_version: LogTime,
+    exceptions: HashSet<LogTime>,
+}
+
+impl VectorEntry {
+    pub fn new() -> Self {
+        VectorEntry::default()
+    }
+
+    pub fn increment(&mut self, clk: LogTime) {
+        match clk.cmp(&self.next_version) {
+            // We've resolved an exception
+            Ordering::Less => {
+                self.exceptions.remove(&clk);
+            }
+            // This is what we expected to see as the next op
+            Ordering::Equal => self.next_version += 1,
+            // We've just found an exception
+            Ordering::Greater => (self.next_version + 1..clk).for_each(|i| {
+                self.exceptions.insert(i);
+            }),
+        };
+    }
+
+    pub fn is_ready(&self, clk: LogTime) -> bool {
+        clk < self.next_version && self.no_exceptions(clk)
+    }
+
+    /// Calculate the difference between a remote VectorEntry and ours.
+    /// Specifically, we want the set of operations we've seen that the remote hasn't
+    pub fn diff_from(&self, other: &Self) -> HashSet<LogTime> {
+        // 1. Find (new) operations that we've seen locally that the remote hasn't
+        let local_ops =
+            (other.next_version..self.next_version).filter(|ix: &LogTime| self.no_exceptions(*ix));
+
+        // 2. Find exceptions that we've seen.
+        let mut local_exceptions = other.exceptions.difference(&self.exceptions).cloned();
+
+        local_ops.chain(&mut local_exceptions).collect()
+    }
+
+    fn no_exceptions(&self, clk: LogTime) -> bool {
+        !self.exceptions.contains(&clk)
+    }
+}
+
+pub trait CausalOp<A> {
+    /// TODO: result should be a VClock<A> since an op could be dependant on a few different msgs
+    /// If the result is Some(dot) then this operation cannot occur until the operation that
+    /// occured at dot has.
+    fn happens_after(&self) -> Option<Dot<A>>;
+
+    /// The time that the current operation occured at
+    fn dot(&self) -> Dot<A>;
+}
+
+impl<A: Actor, T: CausalOp<A>> CausalityBarrier<A, T> {
+    pub fn new() -> Self {
+        CausalityBarrier {
+            peers: HashMap::new(),
+            buffer: HashMap::new(),
+        }
+    }
+
+    pub fn ingest(&mut self, op: T) -> Option<T> {
+        let v = self.peers.entry(op.dot().actor).or_default();
+        // Have we already seen this op?
+        if v.is_ready(op.dot().counter) {
+            return None;
+        }
+
+        v.increment(op.dot().counter);
```

Comment:
I feel like we should be incrementing only after releasing the op. Otherwise we risk dropping ops from a source if we receive ops out of order.

Ie. imagine we receive two ops from site A, but we ingest the op with the more recent time before the older op. In this case, the older op will be dropped

Reply:
To clarify:

Assume we have:

OP1: (A, Clock1)
OP2: (A, Clock2)

with Clock2 < Clock1

Let's assume we have an empty VVwE
We process OP1 first. 

Our VVwE state will be { A: { next_version: Clock1 + 1, exceptions: [0 ... Clock1) } }

When we see OP2, the operation will not be 'ready' because it will be in the exception set. 

Assuming it doesn't depend on anything then we'll produce `Some(op)` on line 124.

Reply:
Ah thanks for expanding, I keep thinking in terms of vclocks, VectorEntry::increment does more than just increment!

Ok perhaps we should move to VectorEntry::ingest to clarify this